src/avram/callbacks.cr

```diff
@@ -1,61 +1,128 @@
         previous_def
       {% end %}
 
-      {{ method.id }}
+      {{ method_name.id }}
     end
   end
 
-  macro after_save(method)
-    def after_save(object)
-      {% if @type.methods.map(&.name).includes?(:after_save.id) %}
+  # Run the given block before saving or creating
+  #
+  # This runs before saving and before the database transaction is started.
+  # You can set defaults, validate, or perform any other setup necessary for
+  # saving.
+  #
+  # ```
+  # before_save do
+  #   validate_required name, age
+  # end
+  # ```
+  macro before_save
+    def before_save
+      {% if @type.methods.map(&.name).includes?(:before_save.id) %}
         previous_def
       {% end %}
 
-      {{ method.id }}(object)
+      {{ yield }}
     end
   end
 
-  macro before_create(method)
-    def before_create
-      {% if @type.methods.map(&.name).includes?(:before_create.id) %}
+  # Run the given method after save, but before transaction is committed
+  #
+  # This is a great place to do other database saves because if something goes
+  # wrong the whole transaction would be rolled back.
+  #
+  # The newly saved record will be passed to the method.
+  #
+  # ```
+  # class SaveComment < Comment::SaveOperation
+  #   after_save touch_post
+  #
+  #   private def touch_post(comment : Comment)
+  #     SavePost.update!(comment.post!, updated_at: Time.utc)
+  #   end
+  # end
+  # ```
+  #
+  # > This is *not* a good place to do things like send messages, enqueue
+  # > background jobs, or charge payments. Since the transaction could be rolled
+  # > back the record may not be persisted to the database.
+  # > Instead use `after_commit`
+  macro after_save(method_name)
+    def after_save(object)
+      {% if @type.methods.map(&.name).includes?(:after_save.id) %}
         previous_def
       {% end %}
 
-      {{ method.id }}
+      {{ method_name.id }}(object)
     end
   end
 
-  macro after_create(method)
-    def after_create(object)
-      {% if @type.methods.map(&.name).includes?(:after_create.id) %}
+  # Run the given method after save and after successful transaction commit
+  #
+  # The newly saved record will be passed to the method.
+  #
+  # ```
+  # class SaveComment < Comment::SaveOperation
+  #   after_commit notify_post_author
+  #
+  #   private def notify_post_author(comment : Comment)
+  #     NewCommentNotificationEmail.new(comment, to: comment.author!).deliver_now
+  #   end
+  # end
+  # ```
+  #
+  macro after_commit(method_name)
+    def after_commit(object)
+      {% if @type.methods.map(&.name).includes?(:after_commit.id) %}
         previous_def
       {% end %}
 
-      {{ method.id }}(object)
+      {{ method_name.id }}(object)
     end
   end
 
-  macro before_update(method)
-    def before_update
-      {% if @type.methods.map(&.name).includes?(:before_update.id) %}
-        previous_def
-      {% end %}
+  {% for callbacks_without_block in [:after_save, :after_commit] %}
```

Comment:
Can't use callbacks with a block because  there is no way to set the block arg.

```crystal
macro  after_save 
  {{ yield }}
end

# Can be called with any number of block args.  They're all ignored
before_save do |foo, bar, baz|
end
```

So we can only allow named methods. I think this is ok since  those  steps typically benefit from a  well named method